### PR TITLE
add reworked chronos job runner

### PR DESCRIPTION
**Pull request by mtangaro**
Dear all,
during the last Europe Biohackathon (2020) we reworked the Galaxy Mesos (Chronos) job runner, allowing it to properly work.
Meanwhile, I also did some other improvements and tests.
This is a joint effort by me, @gmauro, @pmandreoli, @maricaantonacci, @aknijn and, of course, @bgruening.

Brief changelog:

1. Fix the working directory inside the container: the working directory in the container was not right, resulting in the impossibility to retrieve the output files.
   
2. GALAXY_SLOTS and GALAXY_MEMORY_MB added to the json and available within the Container.
   GALAXY_SLOTS set to docker_cpu
   GALAXY_MEMORY_MB set to docker_memory
It has been tested using this [documentation page](https://galaxyproject.org/admin/config/galaxy_slots/).
   
3. A new command script, named chronos_{{job number}}.sh is created instead passing the command in the container preventing wrong characters parsing.
This solve the following issue: https://github.com/galaxyproject/galaxy/issues/10677
Not sure here if this is the right way to implement this, since I'm not using a template and it exploits the $SHELL variable to run it. This variable is correctly handled with biocontainers and, for example, ubuntu images. Any suggestion is more than welcome.
   
4. Now there is the possibility to mount multiple path, e.g. galaxy database path and CVMFS reference data, in the Container, with proper permissions.

5. I added the fail_job function to the runner to correctly show tools stderr in Galaxy. The stderr and stdout are correctly saved in the corresponding file under “outputs”. But, in case of job failure, during the “stop_job” step DB entry with the stderr is overwritten (with nothing).
To fix this, basically, I call the “self._finish_or_resubmit_job” after the “job_state.job_wrapper.fail”, otherwise the last one overwrite the stderr entry in the DB. Please let me know if there is a better way to do this.

6. Minor fix to job_conf.xml comments and dependencies list.

@gmauro and @maricaantonacci also setup a docker compose [here](https://github.com/gmauro/docker-galaxy-stable/blob/bba85c2b65b289c7e11a3536d7ccc208135e923b/compose/docker-compose.mesos.yml#L9) to have it continuously tested. I hope we can do soon a PR for it.

@pmandreoli tested it with the following tools: bowtie2, trimmomatic, trigalore, sort, fastqc, using the galaxyproject CVMFS for the reference data.

To Do list:
this is a first working version, there is still room for improvements. For example, in case of many retries of the same job, the "stop_job" step is not correctly handled.

Marco

### CODE REVIEW

**Comment by gmauro** [comment]
Super!

**Comment by gmauro** [comment]
I created a PR for the docker-compose setup [here](https://github.com/bgruening/docker-galaxy-stable/pull/574)

**Comment by jmchilton** [comment]
Can you fix the linting issues here:

```
./lib/galaxy/jobs/runners/chronos.py:65:5: E731 do not assign a lambda expression, use a def
./lib/galaxy/jobs/runners/chronos.py:65:20: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:65:22: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:67:14: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:67:20: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:68:42: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:68:48: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:72:43: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:73:3: E114 indentation is not a multiple of four (comment)
./lib/galaxy/jobs/runners/chronos.py:74:3: E114 indentation is not a multiple of four (comment)
./lib/galaxy/jobs/runners/chronos.py:75:3: E114 indentation is not a multiple of four (comment)
./lib/galaxy/jobs/runners/chronos.py:76:3: E731 do not assign a lambda expression, use a def
./lib/galaxy/jobs/runners/chronos.py:76:3: E111 indentation is not a multiple of four
./lib/galaxy/jobs/runners/chronos.py:76:21: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:76:26: E201 whitespace after '{'
./lib/galaxy/jobs/runners/chronos.py:76:48: E202 whitespace before '}'
./lib/galaxy/jobs/runners/chronos.py:77:3: E111 indentation is not a multiple of four
./lib/galaxy/jobs/runners/chronos.py:77:7: E222 multiple spaces after operator
./lib/galaxy/jobs/runners/chronos.py:77:10: E201 whitespace after '['
./lib/galaxy/jobs/runners/chronos.py:77:77: E202 whitespace before ']'
./lib/galaxy/jobs/runners/chronos.py:78:3: E111 indentation is not a multiple of four
./lib/galaxy/jobs/runners/chronos.py:80:1: E302 expected 2 blank lines, found 1
./lib/galaxy/jobs/runners/chronos.py:212:51: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:212:54: E231 missing whitespace after ','
./lib/galaxy/jobs/runners/chronos.py:217:5: E303 too many blank lines (2)
./lib/galaxy/jobs/runners/chronos.py:232:19: E111 indentation is not a multiple of four
./lib/galaxy/jobs/runners/chronos.py:234:19: E111 indentation is not a multiple of four
./lib/galaxy/jobs/runners/chronos.py:280:5: E115 expected an indented block (comment)
./lib/galaxy/jobs/runners/chronos.py:281:5: E115 expected an indented block (comment)
./lib/galaxy/jobs/runners/chronos.py:286:13: E225 missing whitespace around operator
./lib/galaxy/jobs/runners/chronos.py:300:13: E265 block comment should start with '# '
./lib/galaxy/jobs/runners/chronos.py:319:105: E202 whitespace before '}'
```

**Comment by mtangaro** [comment]
Hi @jmchilton, I've fixed the linting issues.
I have two more issues, which actually I had not before.
The one from "Integration / Test (3.7, kubernetes" seems not related to the runner.
While I can't see any log from the "get_code_and_test" test.
Can you have a look?

**Comment by nsoranzo** [comment]
@mtangaro The kubernetes issue should be fixed if you rebase or merge `dev`.

**Comment by mtangaro** [comment]
Thanks a lot @nsoranzo. Everything is ok now :)

File: lib/galaxy/config/sample/job_conf.xml.sample_advanced [hunk 1]
```diff
@@ -1061,12 +1061,12 @@
...
-            <param id="volumes">/directory/</param>
+            <param id="volumes">host_path:container_path:mode</param>
             <!-- Directory which is mounted to the container and is parent of
                  the  `job_working_directory`, `file_path`, `new_file_path`
-                 directories. Directories of the data used by tools are
-                 included as well.-->
-            <param id="max_retries">2</param>
+                 directories. Other directories of the data used by tools can
+                 be mounted as well, separated by commas.-->
+            <param id="max_retries">0</param>
```
Comment (jmchilton) [inline lib/galaxy/config/sample/job_conf.xml.sample_advanced:1069]: Any chance I can get you to update the corresponding docs in the sample YAML file (https://github.com/galaxyproject/galaxy/blob/dev/test/unit/jobs/job_conf.sample_advanced.yml). It copied a lot of the comments out of this sample XML but I haven't gotten around to replacing the XML version with the YAML version - but the YAML will be the go to at some point soon.

Comment (mtangaro) [reply lib/galaxy/config/sample/job_conf.xml.sample_advanced:1069]: This goes without saying! Next time I will update it by default.

**Comment by jmchilton** [comment]
Thanks so much!

**Pull request merged**